test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,229 @@
+import '../example_data.dart' as eg;
+
+Map<int, Map<int, Set<int>>> _streamTrackerToSet(Map<int, Map<int, MessageIdTracker>> streamSenders) {
+  final Map<int, Map<int, Set<int>>> messagesByUserInStream = {};
+  for (final entry in streamSenders.entries) {
+    final MapEntry(key: streamId, value: sendersByStream) = entry;
+    for (final entry in sendersByStream.entries) {
+      final MapEntry(key: senderId, value: tracker) = entry;
+      (messagesByUserInStream[streamId] ??= {})[senderId] = Set.from(tracker.ids);
+    }
+  }
+  return messagesByUserInStream;
+}
+
+Map<int, Map<String, Map<int, Set<int>>>> _topicTrackerToSet(
+  Map<int, Map<String, Map<int, MessageIdTracker>>> topicSenders) {
+    Map<int, Map<String, Map<int, Set<int>>>> messagesByUserInTopic = {};
+    for (final entry in topicSenders.entries) {
+      final MapEntry(key: streamId, value: topicsByStream) = entry;
+      for (final entry in topicsByStream.entries) {
+        final MapEntry(key: topic, value: sendersByTopic) = entry;
+        for (final entry in sendersByTopic.entries) {
+          final MapEntry(key: senderId, value: tracker) = entry;
+          ((messagesByUserInTopic[streamId] ??= {})[topic] ??= {})
+            [senderId] = Set.from(tracker.ids);
+        }
+      }
+    }
+    return messagesByUserInTopic;
+}
+
+/// [messages] should be sorted by [id] ascendingly.
+void checkMatchesMessages(RecentSenders model, List<Message> messages) {
+  final Map<int, Map<int, Set<int>>> messagesByUserInStream = {};
+  final Map<int, Map<String, Map<int, Set<int>>>> messagesByUserInTopic = {};
+  for (final message in messages) {
+    if (message is! StreamMessage) {
+      throw UnsupportedError('Message of type ${message.runtimeType} is not expected.');
+    }
+
+    final StreamMessage(:streamId, :topic, :senderId, id: int messageId) = message;
+
+    ((messagesByUserInStream[streamId] ??= {})
+      [senderId] ??= {}).add(messageId);
+    (((messagesByUserInTopic[streamId] ??= {})[topic] ??= {})
+      [senderId] ??= {}).add(messageId);
+  }
+  final actualMessagesByUserInStream = _streamTrackerToSet(model.streamSenders);
+  final actualMessagesByUserInTopic = _topicTrackerToSet(model.topicSenders);
+  check(actualMessagesByUserInStream).deepEquals(messagesByUserInStream);
+  check(actualMessagesByUserInTopic).deepEquals(messagesByUserInTopic);
+}
+
+void main() {
+  group('RecentSenders', () {
+    test('starts with empty stream and topic senders', () {
+      final model = RecentSenders();
+      checkMatchesMessages(model, []);
+    });
+
+    test('stream message gets included, DM message gets ignored', () {
+      final model = RecentSenders();
+      final streamMessage = eg.streamMessage();
+      final dmMessage = eg.dmMessage(from: eg.selfUser, to: [eg.otherUser]);
+
+      model.handleMessage(streamMessage);
+      model.handleMessage(dmMessage);
+      checkMatchesMessages(model, [streamMessage]);
+    });
+
+    group('add a batch of messages', () {
```

Comment:
nit: name tests after the code's identifiers for what they're testing (here and a few other places):
```suggestion
    group('handleMessages', () {
```

That helps make it easy to navigate from a piece of the app's code to its tests and back.

Reply:
Better yet: call this `RecentSenders.handleMessages`, and remove the outer `group` layer.

There's no real need for a `RecentSenders` group around the outside — this whole file is for testing `RecentSenders`, and given it's called `recent_senders_test.dart` that can be expected to continue :slightly_smiling_face: